v2/relay/constraints.go

```diff
@@ -0,0 +1,187 @@
+package relay
+
+import (
+	crand "crypto/rand"
+	"encoding/binary"
+	"errors"
+	"math/rand"
+	"sync"
+	"time"
+
+	asnutil "github.com/libp2p/go-libp2p-asn-util"
+	"github.com/libp2p/go-libp2p-core/peer"
+	ma "github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr/net"
+)
+
+var cleanupInterval = 2 * time.Minute
+var validity = 30 * time.Minute
+
+var (
+	errTooManyReservations        = errors.New("too many reservations")
+	errTooManyReservationsForPeer = errors.New("too many reservations for peer")
+	errTooManyReservationsForIP   = errors.New("too many peers for IP address")
+	errTooManyReservationsForASN  = errors.New("too many peers for ASN")
+)
+
+// constraints implements various reservation constraints
+type constraints struct {
+	rc *Resources
+
+	closed                  bool
+	closing, cleanupRunning chan struct{}
+
+	mutex sync.Mutex
+	rand  rand.Rand
+	total map[uint64]time.Time
+	peers map[peer.ID]map[uint64]time.Time
+	ips   map[string]map[uint64]time.Time
+	asns  map[string]map[uint64]time.Time
+}
+
+// NewConstraints creates a new constraints object.
+// The methods are *not* thread-safe; an external lock must be held if synchronization
+// is required.
+func NewConstraints(rc *Resources) *constraints {
+	b := make([]byte, 8)
+	crand.Read(b)
+	random := rand.New(rand.NewSource(int64(binary.BigEndian.Uint64(b))))
+
+	c := &constraints{
+		rc:             rc,
+		closing:        make(chan struct{}),
+		cleanupRunning: make(chan struct{}),
+		rand:           *random,
+		total:          make(map[uint64]time.Time),
+		peers:          make(map[peer.ID]map[uint64]time.Time),
+		ips:            make(map[string]map[uint64]time.Time),
+		asns:           make(map[string]map[uint64]time.Time),
+	}
+	go c.cleanup()
+	return c
+}
+
+// AddReservation adds a reservation for a given peer with a given multiaddr.
+// If adding this reservation violates IP constraints, an error is returned.
+func (c *constraints) AddReservation(p peer.ID, a ma.Multiaddr) error {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if len(c.total) >= c.rc.MaxReservations {
+		return errTooManyReservations
+	}
+
+	ip, err := manet.ToIP(a)
+	if err != nil {
+		return errors.New("no IP address associated with peer")
+	}
+
+	peerReservations := c.peers[p]
+	if len(peerReservations) >= c.rc.MaxReservationsPerPeer {
+		return errTooManyReservationsForPeer
+	}
+
+	ipStr := ip.String()
+	ipReservations := c.ips[ipStr]
+	if len(ipReservations) >= c.rc.MaxReservationsPerIP {
+		return errTooManyReservationsForIP
+	}
+
+	var ansReservations map[uint64]time.Time
+	var asn string
+	if ip.To4() == nil {
+		asn, _ = asnutil.Store.AsnForIPv6(ip)
+		if asn != "" {
+			ansReservations = c.asns[asn]
+			if len(ansReservations) >= c.rc.MaxReservationsPerASN {
+				return errTooManyReservationsForASN
+			}
+		}
+	}
+
+	now := time.Now()
+	id := c.rand.Uint64()
```

Comment:
If we need unique IDs, `new(struct{})` should do it (I think, unless you need `new(byte)`?).

Otherwise, I'd just make a `reservation` struct:

```go
type reservation {
    time time.Time
    peer peer.ID
}
```

And make all of these maps `*reservation` sets (`map[*reservation]struct{}`).

Reply:
Actually, no, there's a _much_ simpler way to do this. Reservations are increasing so we don't even need a map. We can just keep a sorted list of times.

1. When reserving, append.
2. When cleaning, pop from the beginning.

If we want to get even faster (likely not necessary given the reservation limit) we could keep a queue of reservations to garbage collect (again, in time sorted order). Each entry in the queue would include:

1. The reservation's expiration.
2. The key in the ASN map (if any).
3. The key in the IP map (if any).
4. etc.

This would let us drop the GC loop entirely, garbage collecting old reservations on demand every time we try to add a new reservation.